### PR TITLE
internal/dag: Add exact path match condition in HttpProxy

### LINKED ISSUE #4069: Exact path match for HTTPProxy
It'd be nice to have exact patch match for HTTPProxy.
HTTPProxy seems to support only prefix path match, even though exact path match is included in [this design document](https://github.com/projectcontour/contour/blob/main/design/routing-design.md#exact-match).
Now, I need to configure a route for exact '/' and a route for prefix '/' at the same time, and I have no idea how to do it with HTTPProxy.

### PR DESCRIPTION
**Pull request by arjunsalyan**
This adds support for exact path matching in the HttpProxy along side prefix matching

fixes #4069
Signed-off-by: Arjun Salyan <iarjunsalyan@gmail.com>

### CODE REVIEW

**Comment by arjunsalyan** [comment]
This is by no means near completion. Still need to cover some edge cases as I feel I must have missed them, and work on the documentation. Any reviews and feedback would be very helpful. Also, please bear with me as this is my first PR.

File: design/httpproxy-status-conditions.md
Comment (sunjayBhatia) [inline design/httpproxy-status-conditions.md:280]: no need to modify old design docs for this change

**Review by sunjayBhatia — COMMENTED** [review]
re the rules around merging path conditions in the tree of HTTPProxy includes:

```
// If the root condition is prefix, and included condition is prefix then after merging we get one prefix condition.
// If the root condition is prefix, and included condition is exact then after merging we get one exact condition.
// If the root condition is exact, and included condition is exact then after merging we get one exact condition.
// If the root condition is exact, and included condition is prefix then after merging we get one prefix condition.
```

an alternate interpretation of the "exact" match in this context could be that such a match is terminal, i.e. it must come as a leaf of the tree of match conditions, since as it stands above if an exact match is at the root or middle of an include tree it effectively becomes a prefix match. if we want to keep these things distinct/clear that might be a way to go

File: internal/dag/httpproxy_processor.go
Comment (sunjayBhatia) [inline internal/dag/httpproxy_processor.go:1425]: the `seenConds` map key will need some changes to differentiate they type of the merged path, since it is valid to have a prefix or an exact match for the same path string

File: internal/dag/conditions.go
Comment (sunjayBhatia) [inline internal/dag/conditions.go:35]: we'll definitely need some thorough unit tests to make sure this logic is well defined and has proper coverage

**Comment by sunjayBhatia** [comment]
Thanks for the PR @arjunsalyan I think this is a good start, but we will definitely need some more unit test coverage, we can give pointers where to look to add these if you like

we should already have logic further in our processing logic to sort the generated routes properly so that exact matches are considered by Envoy first, so this should mostly be covered by httpproxy processor/dag tests and some end to end tests

File: internal/dag/conditions.go
Comment (arjunsalyan) [reply internal/dag/conditions.go:35]: > re the rules around merging path conditions in the tree of HTTPProxy includes:
> 
> ```
> // If the root condition is prefix, and included condition is prefix then after merging we get one prefix condition.
> // If the root condition is prefix, and included condition is exact then after merging we get one exact condition.
> // If the root condition is exact, and included condition is exact then after merging we get one exact condition.
> // If the root condition is exact, and included condition is prefix then after merging we get one prefix condition.
> ```
> 
> an alternate interpretation of the "exact" match in this context could be that such a match is terminal, i.e. it must come as a leaf of the tree of match conditions, since as it stands above if an exact match is at the root or middle of an include tree it effectively becomes a prefix match. if we want to keep these things distinct/clear that might be a way to go

Agreed @sunjayBhatia, If you see the next line ahead of these comments I have mentioned the interpretation. These comments are just to explain why we reached the conclusion that the leaf decides the path type.

```
// Hence this can be concluded that the last rule in the delegation chain decides the path type
```

Let me know if you meant something else.

**Comment by arjunsalyan** [comment]
> re the rules around merging path conditions in the tree of HTTPProxy includes:
> 
> ```
> // If the root condition is prefix, and included condition is prefix then after merging we get one prefix condition.
> // If the root condition is prefix, and included condition is exact then after merging we get one exact condition.
> // If the root condition is exact, and included condition is exact then after merging we get one exact condition.
> // If the root condition is exact, and included condition is prefix then after merging we get one prefix condition.
> ```
> 
> an alternate interpretation of the "exact" match in this context could be that such a match is terminal, i.e. it must come as a leaf of the tree of match conditions, since as it stands above if an exact match is at the root or middle of an include tree it effectively becomes a prefix match. if we want to keep these things distinct/clear that might be a way to go

Agreed @sunjayBhatia, If you see the next line ahead of these comments I have mentioned the interpretation. These comments are just to explain why we reached the conclusion that the leaf decides the path type.

```
// Hence this can be concluded that the last rule in the delegation chain decides the path type
```

Let me know if you meant something else.

**Comment by sunjayBhatia** [comment]
> > re the rules around merging path conditions in the tree of HTTPProxy includes:
> > ```
> > // If the root condition is prefix, and included condition is prefix then after merging we get one prefix condition.
> > // If the root condition is prefix, and included condition is exact then after merging we get one exact condition.
> > // If the root condition is exact, and included condition is exact then after merging we get one exact condition.
> > // If the root condition is exact, and included condition is prefix then after merging we get one prefix condition.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > an alternate interpretation of the "exact" match in this context could be that such a match is terminal, i.e. it must come as a leaf of the tree of match conditions, since as it stands above if an exact match is at the root or middle of an include tree it effectively becomes a prefix match. if we want to keep these things distinct/clear that might be a way to go
> 
> Agreed @sunjayBhatia, If you see the next line ahead of these comments I have mentioned the interpretation. These comments are just to explain why we reached the conclusion that the leaf decides the path type.
> 
> ```
> // Hence this can be concluded that the last rule in the delegation chain decides the path type
> ```
> 
> Will improve the wording of this statement. Let me know if you meant something else.

Above I meant that an option we can take is to not allow and mark an HTTPProxy as Invalid if you try to use an exact match in an include match condition

**Comment by arjunsalyan** [comment]
> Above I meant that an option we can take is to not allow and mark an HTTPProxy as Invalid if you try to use an exact match in an include match condition

Got it. I have some subjective thoughts here, especially for the case of delegation of a path to teams. Included routes should have the flexibility to use exact matches. But open to any other views on this.

I have added unit tests and e2e trying to cover most of the things. Will start with documentation once these things come to good shape.

**Comment by arjunsalyan** [comment]
For now I have added the docs considering my initial theory about routes delegation, just to make this PR ready for a review. Although we can revisit it.

**Review by sunjayBhatia — COMMENTED** [review]
I think my interpretation of "Exact" path matching in conjunction with the Inclusion feature is that if you are using Inclusion, you really want a "Prefix" match (because the Included child proxy can then do different matching) so "Exact" matches should be disallowed on the path match conditions on an Include

cc @skriss @tsaarni if you have any thoughts

**Comment by arjunsalyan** [comment]
> > I think my interpretation of "Exact" path matching in conjunction with the Inclusion feature is that if you are using Inclusion, you really want a "Prefix" match (because the Included child proxy can then do different matching) so "Exact" matches should be disallowed on the path match conditions on an Include
> > cc @skriss @tsaarni if you have any thoughts
> 
> Yeah, that is how I had been thinking about it too, should be disallowed as an Include condition.

Alright, I am starting to see the point here. But before we reach a conclusion, please allow me to put one last statement in the favour of the proposed merging algorithm.

- Since we have a strict rule of one fqdn - one httpproxy, inclusion becomes the only way to route traffic to services in different namespaces from the same fqdn. To adopt a general approach, we currently follow a methodology where we create an orphan httpproxy without any rules in all of the namespaces. Then we create a root httpproxy in one of the namespaces and define all the rules in that root httpproxy. Not having the exact condition in include takes away this flexibility.

There are definitely other ways to handle this, but just putting a use case.

Anyways, I am now aligned on not allowing the `Exact` in the Include condition. Will push the changes soon.

**Comment by arjunsalyan** [comment]
Hi @skriss @sunjayBhatia 

The changes for disallowing the `exact` condition except for the child/leaf httproxy are done.

File: site/content/docs/main/config/request-routing.md
Comment (sunjayBhatia) [inline site/content/docs/main/config/request-routing.md:94]:
```suggestion
only allowed in route match conditions and not in include match conditions.
```

File: site/content/docs/main/config/inclusion-delegation.md
Comment (sunjayBhatia) [inline site/content/docs/main/config/inclusion-delegation.md:29]:
```suggestion
- `exact:` conditions are also concatenated just like `prefix:` conditions, but `exact:` conditions are not allowed in include match conditions. If the child httpproxy has `exact:` condition then after concatenation, it becomes a single `exact:` condition. For example, `prefix: /static` and `exact: /main.js` become a single `exact: /static/main.js` condition.
```

File: apis/projectcontour/v1/httpproxy.go
Comment (sunjayBhatia) [inline apis/projectcontour/v1/httpproxy.go:70]: would be good to note here that this field cannot be used in an include match condition

File: internal/dag/httpproxy_processor.go
Comment (sunjayBhatia) [inline internal/dag/httpproxy_processor.go:628]: we should add a test in `status_test.go` for this error case

File: internal/dag/httpproxy_processor.go
Comment (sunjayBhatia) [inline internal/dag/httpproxy_processor.go:627]: we should add a test in `builder_test.go` for the changes in this file, there should be example tests in there for prefix matching that can be used as a template

Comment (sunjayBhatia) [reply internal/dag/httpproxy_processor.go:627]: in general since codespell/codegen have failed, we aren't seeing the coverage report which I think will show we need some more unit tests for changes in this file

File: internal/dag/httpproxy_processor.go
Comment (sunjayBhatia) [inline internal/dag/httpproxy_processor.go:1427]: we should add some unit testing to `TestIncludeMatchConditionsIdentical` for the changes in this function

Comment (arjunsalyan) [reply internal/dag/httpproxy_processor.go:1427]: I was wondering if this function even needs to be changed now. Because include conditions will only have prefix conditions and not exact conditions, I don't see a need to change this function. But since `mergePathMatchConditions` has the possibility to return an `ExactMatchCondition`, so we should just add something to handle that part.

Comment (sunjayBhatia) [reply internal/dag/httpproxy_processor.go:1427]: I think we may be able to rely on the fact that include match conditions should never have exact path matches in them, validation should happen before this is called so defo fair point there shouldn't really be changes to this

**Comment by arjunsalyan** [comment]
@sunjayBhatia Added the generated sources, which were leading to the failed test.

**Comment by arjunsalyan** [comment]
This is strange

```
Build and Test Pull Request / test-linux (pull_request)
Build and Test Pull Request / test-osx (pull_request)
```

These tests pass all fine when I run them locally. Looking into it

**Comment by sunjayBhatia** [comment]
> This is strange
> 
> ```
> Build and Test Pull Request / test-linux (pull_request)
> Build and Test Pull Request / test-osx (pull_request)
> ```
> 
> These tests pass all fine when I run them locally. Looking into it

you'll want to merge/rebase on main, and update the port in that test (github is merging the PR with the base branch when checking out the code)

**Comment by arjunsalyan** [comment]
> you'll want to merge/rebase on main, and update the port in that test (github is merging the PR with the base branch when checking out the code)

Ah! The tests were updated. Fixed

File: test/e2e/httpproxy/exact_path_condition_match_test.go
Comment (sunjayBhatia) [inline test/e2e/httpproxy/exact_path_condition_match_test.go:132]: nit: I think this looks correct, but it's not super intuitive to read

not a huge deal but would be nice to either add some comments above some of the cases or use naming of the routes/services that make it clearer

File: test/e2e/httpproxy/include_exact_condition_test.go
Comment (sunjayBhatia) [inline test/e2e/httpproxy/include_exact_condition_test.go:191]: similar nit to the above on comments/naming to make these test cases super clear (looks correct though)

File: internal/dag/builder_test.go
Comment (sunjayBhatia) [inline internal/dag/builder_test.go:15093]: nit: the `prefixString` was named since it represented a prefix match that would basically match a path using "string prefix" vs. matching path segments, we could change this to just `exact` here

**Review by sunjayBhatia — APPROVED** [review]
looks good overall, just some tiny nits

approving but will leave for another maintainer to have a pass over as well before merge

File: internal/dag/builder_test.go
Comment (arjunsalyan) [reply internal/dag/builder_test.go]: Makes sense, rather `exact()` is already there making it completely redundant.

File: test/e2e/httpproxy/exact_path_condition_match_test.go
Comment (arjunsalyan) [reply test/e2e/httpproxy/exact_path_condition_match_test.go:132]: Can agree to that, it is just the nature of a few cases that I could not find better naming. Have tried something with comments, not sure how much value that adds though.

**Review by skriss — APPROVED** [review]
LGTM, thanks @arjunsalyan!

**Pull request merged**
